src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/MultipartUploader.java

```diff
@@ -0,0 +1,256 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.sonatype.nexus.blobstore.api.BlobStoreException;
+import org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobTargetOption;
+import com.google.cloud.storage.Storage.ComposeRequest;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Component that provides parallel multipart upload support for blob binary data (.bytes files).
+ */
+@Named
+public class MultipartUploader
+    extends StateGuardLifecycleSupport
+{
+
+  /**
+   * Use this property in 'nexus.properties' to control how large each multipart part is. Default is 5 MB.
+   * Smaller numbers increase the number of parallel workers used to upload a file. Match to your workload:
+   * if you are heavy in docker with large images, increase; if you are heavy in smaller components, decrease.
+   */
+  public static final String CHUNK_SIZE_PROPERTY = "nexus.gcs.multipartupload.chunksize";
+
+  /**
+   * This is a hard limit on the number of components to a compose request enforced by Google Cloud Storage API.
+   */
+  static final int COMPOSE_REQUEST_LIMIT = 32;
+
+  /**
+   * While an invocation of {@link #upload(Storage, String, String, InputStream)} is in-flight, the individual
+   * chunks of the file will have names like 'destination.chunkPartNumber", like
+   * 'content/vol-01/chap-01/UUID.bytes.chunk1', 'content/vol-01/chap-01/UUID.bytes.chunk2', etc.
+   */
+  private final String CHUNK_NAME_PART = ".chunk";
+
+  /**
+   * Used internally to count how many times we've hit the compose limit.
+   * Consider exposing this as a bean that can provide tuning feedback to deployers.
+   */
+  private final AtomicLong composeLimitHit = new AtomicLong(0);
+
+  private static final byte[] EMPTY = new byte[0];
+
+  private final ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(
+          new ThreadFactoryBuilder().setNameFormat("nexus-google-cloud-storage-multipart-upload-%d")
+              .build()));
+
+  private final int chunkSize;
+
+  @Inject
+  public MultipartUploader(@Named("${"+CHUNK_SIZE_PROPERTY +":-5242880}") final int chunkSize) {
+    this.chunkSize = chunkSize;
+  }
+
+  @Override
+  protected void doStop() throws Exception {
+    executorService.shutdown();
```

Comment:
Suggestion: Logging at info feels a bit spammy.  Maybe log at debug and log something at info if `awaitTermination` times out?

Reply:
> Suggestion: Logging at info feels a bit spammy. Maybe log at debug and log something at info if awaitTermination times out?

Glad you pointed this out. It felt weird to wait a few minutes on orderly shutdown for any uploading threads to complete. Now that I think about it, there may be other services stopping that may not be available when these uploads complete, making the partial uploads useless.

I was looking at the [S3 blobstore](https://github.com/sonatype/nexus-public/blob/master/plugins/nexus-blobstore-s3/src/main/java/org/sonatype/nexus/blobstore/s3/internal/ParallelRequester.java#L72), and it looks like it just executes a `shutdownNow()` without waiting. I'm thinking I should follow suit - wdyt?

Reply:
In e394e051 I went ahead and followed the S3 Blobstore approach, so the questionable log statement is now gone.